In [84]:
import pandas as pd

In [178]:
df = pd.read_excel('/Users/agustindei/Documents/Jedha/Full_Stack/Projets/projet_deploiement/get_around_delay_analysis.xlsx')
df.head()

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0     505000  363965       mobile  canceled                           NaN   
1     507750  269550       mobile     ended                         -81.0   
2     508131  359049      connect     ended                          70.0   
3     508865  299063      connect  canceled                           NaN   
4     511440  313932       mobile     ended                           NaN   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                       NaN                                         NaN  
1                       NaN                                         NaN  
2                       NaN                                         NaN  
3                       NaN                                         NaN  
4                       NaN                                         NaN

In [173]:
df.shape

(21310, 7)

In [87]:
# Basic stats
print("Number of rows : {}".format(df.shape[0]))
print()

print("Display of dataset: ")
display(df.head())
print()

print("Basics statistics: ")
data_desc = df.describe(include="all")
display(data_desc)
print()

print("Percentage of missing values: ")
display(100 * df.isnull().sum() / df.shape[0])

Number of rows : 21310

Display of dataset: 


rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0     505000  363965       mobile  canceled                           NaN   
1     507750  269550       mobile     ended                         -81.0   
2     508131  359049      connect     ended                          70.0   
3     508865  299063      connect  canceled                           NaN   
4     511440  313932       mobile     ended                           NaN   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                       NaN                                         NaN  
1                       NaN                                         NaN  
2                       NaN                                         NaN  
3                       NaN                                         NaN  
4                       NaN                                         NaN


Basics statistics: 


rental_id         car_id checkin_type  state  \
count    21310.000000   21310.000000        21310  21310   
unique            NaN            NaN            2      2   
top               NaN            NaN       mobile  ended   
freq              NaN            NaN        17003  18045   
mean    549712.880338  350030.603426          NaN    NaN   
std      13863.446964   58206.249765          NaN    NaN   
min     504806.000000  159250.000000          NaN    NaN   
25%     540613.250000  317639.000000          NaN    NaN   
50%     550350.000000  368717.000000          NaN    NaN   
75%     560468.500000  394928.000000          NaN    NaN   
max     576401.000000  417675.000000          NaN    NaN   

        delay_at_checkout_in_minutes  previous_ended_rental_id  \
count                   16346.000000               1841.000000   
unique                           NaN                       NaN   
top                              NaN                       NaN   
freq                             NaN                       NaN   
mean                       59.701517             550127.411733   
std                      1002.561635              13184.023111   
min                    -22433.000000             505628.000000   
25%                       -36.000000             540896.000000   
50%                         9.000000             550567.000000   
75%                        67.000000             560823.000000   
max                     71084.000000             575053.000000   

        time_delta_with_previous_rental_in_minutes  
count                                  1841.000000  
unique                                         NaN  
top                                            NaN  
freq                                           NaN  
mean                                    279.288430  
std                                     254.594486  
min                                       0.000000  
25%                                      60.000000  
50%                                     180.000000  
75%                                     540.000000  
max                                     720.000000


Percentage of missing values: 


rental_id                                      0.000000
car_id                                         0.000000
checkin_type                                   0.000000
state                                          0.000000
delay_at_checkout_in_minutes                  23.294228
previous_ended_rental_id                      91.360863
time_delta_with_previous_rental_in_minutes    91.360863
dtype: float64

Share of the owner’s revenue would potentially be affected by the feature

In [109]:
num_rentals_concerned = df.previous_ended_rental_id.notnull().sum()
print(f"Number of rentals concerned by the new features: ", num_rentals_concerned)

print()
print("Percentage rentals over the total of rentals that can be affected by the new features: ")
percentage_rentals_affected = 100 * df.previous_ended_rental_id.notnull().sum() / df.shape[0]
print(percentage_rentals_affected)

Number of rentals concerned by the new features:  1841

Percentage rentals over the total of rentals that can be affected by the new features: 
8.639136555607696


In [110]:
import plotly.graph_objects as go

# Assuming percentage_rentals_affected has been computed
percentage_rentals_affected = 100 * num_rentals_concerned / df.shape[0]

# Create the bar chart
fig = go.Figure(
    go.Bar(
        x=['Affected Rentals'], 
        y=[percentage_rentals_affected],
        marker_color='skyblue',
        text=[f'{percentage_rentals_affected:.2f}%'],
        textposition='auto'
    )
)

# Customize the layout
fig.update_layout(
    title='Percentage of Rentals Concerned',
    yaxis_title='Percentage (%)',
    yaxis_range=[0, 100]
)

# Display the chart
fig.show()



The share of our owner’s revenue would potentially be affected by the feature is 8,6%

How many rentals would be affected by the feature depending on the threshold and scope we choose?

In [111]:
df_noprob = df[df.previous_ended_rental_id.isnull()]

In [112]:
df.state.value_counts()

state
ended       18045
canceled     3265
Name: count, dtype: int64

In [113]:
df_noprob.state.value_counts()

state
ended       16433
canceled     3036
Name: count, dtype: int64

In [114]:
df_cancelled = df[df['state']=='canceled']

In [115]:
df_cancelled.delay_at_checkout_in_minutes.value_counts()

delay_at_checkout_in_minutes
-17468.0    1
Name: count, dtype: int64

In [116]:
df_cancelled.describe()

rental_id         car_id  delay_at_checkout_in_minutes  \
count    3265.000000    3265.000000                           1.0   
mean   548637.683920  350585.309954                      -17468.0   
std     14907.810897   57254.052866                           NaN   
min    504871.000000  159533.000000                      -17468.0   
25%    539183.000000  317572.000000                      -17468.0   
50%    549700.000000  368593.000000                      -17468.0   
75%    560563.000000  394869.000000                      -17468.0   
max    576195.000000  416935.000000                      -17468.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
count                229.000000                                  229.000000  
mean              550913.327511                                  294.890830  
std                11955.397600                                  250.591601  
min               509972.000000                                    0.000000  
25%               543706.000000                                   60.000000  
50%               550970.000000                                  210.000000  
75%               560395.000000                                  570.000000  
max               574540.000000                                  720.000000

In [117]:
df_cancelled.delay_at_checkout_in_minutes.notnull().sum()

1

In [118]:
df_cancelled.delay_at_checkout_in_minutes.describe()

count        1.0
mean    -17468.0
std          NaN
min     -17468.0
25%     -17468.0
50%     -17468.0
75%     -17468.0
max     -17468.0
Name: delay_at_checkout_in_minutes, dtype: float64

In [119]:
df_cancelled[df_cancelled['delay_at_checkout_in_minutes'] == -17468]

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
21002     559126  379544       mobile  canceled                      -17468.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
21002                       NaN                                         NaN

In [120]:
df_cancelled[df_cancelled['time_delta_with_previous_rental_in_minutes'] == 600]

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
2446      546387  352528      connect  canceled                           NaN   
6771      568659  317378      connect  canceled                           NaN   
8134      528949  383144      connect  canceled                           NaN   
8871      539164  364464      connect  canceled                           NaN   
9393      542034  280406       mobile  canceled                           NaN   
9523      535520  374169      connect  canceled                           NaN   
11073     562896  376677      connect  canceled                           NaN   
13107     536440  392633       mobile  canceled                           NaN   
13178     549066  409663       mobile  canceled                           NaN   
13934     570917  411523       mobile  canceled                           NaN   
14183     556983  338969      connect  canceled                           NaN   
14426     536171  392368      connect  canceled                           NaN   
15700     543223  376311      connect  canceled                           NaN   
17151     564664  361789       mobile  canceled                           NaN   
17870     568519  295783      connect  canceled                           NaN   
18343     538543  330013       mobile  canceled                           NaN   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
2446                   546578.0                                       600.0  
6771                   566412.0                                       600.0  
8134                   540254.0                                       600.0  
8871                   537569.0                                       600.0  
9393                   540818.0                                       600.0  
9523                   564202.0                                       600.0  
11073                  563347.0                                       600.0  
13107                  553304.0                                       600.0  
13178                  544279.0                                       600.0  
13934                  568349.0                                       600.0  
14183                  552161.0                                       600.0  
14426                  515799.0                                       600.0  
15700                  560879.0                                       600.0  
17151                  564691.0                                       600.0  
17870                  561162.0                                       600.0  
18343                  540896.0                                       600.0

## Drivers on time vs drivers late for check-out

In [145]:

drivers_on_time = (100 * (df['delay_at_checkout_in_minutes'] <= 0).sum() / df.shape[0])
print("Percentage of drivers who returned their car on time or before the scheduled time: ")
print(drivers_on_time)

Percentage of drivers who returned their car on time or before the scheduled time: 
32.57625527921164


In [146]:
drivers_late = (df['delay_at_checkout_in_minutes'] > 0).sum()
print(f"Number of drivers late for check-out: ", drivers_late)

print()
percentage_drivers_late = (100 * (df['delay_at_checkout_in_minutes'] > 0).sum() / df.shape[0])
print("Percentage of drivers late for check-out: ")
print(percentage_drivers_late)

Number of drivers late for check-out:  9404

Percentage of drivers late for check-out: 
44.12951665884561


There are more drivers who are late for the check-out than drivers on time:

32,5% of drivers returned their car on time or before the scheduled time.
44% of drivers are late for check-out.

How does it impact the next driver?


44 % of drivers are late for check-out and therefore the cars are late for the next check-in. 

When this happens, the minimum waiting time was 30 minutes and 720 minutes in worst case scenario. 75% of the clients, who received a car handed-in with a delay in the check-out, waited up to 600 minutes, 50% of them waited up to 240 minutes. On average this group of clients had to wait 329 minutes, but with a standard deviation of 244 minutes. 

The dataset doesn't show any data when the time difference between two rentals is more than 12 hours. 

In [147]:
import plotly.graph_objects as go


# Data for the bar chart
labels = ['Drivers Late', 'Drivers On Time']
values = [percentage_drivers_late, drivers_on_time]

# Create the bar chart
fig = go.Figure(
    go.Bar(
        x=labels, 
        y=values,
        marker_color=['skyblue', 'lightcoral', 'lightgreen'],
        text=[f'{val:.2f}%' for val in values],
        textposition='auto'
    )
)

# Customize the layout
fig.update_layout(
    title='Percentage of Rentals Concerned and Drivers Status',
    yaxis_title='Percentage (%)',
    yaxis_range=[0, 100]
)

# Display the chart
fig.show()


In [124]:
df.time_delta_with_previous_rental_in_minutes.notnull().sum()

1841

In [148]:
df.previous_ended_rental_id.notnull().sum()

1841

In [183]:
df.shape

(21310, 7)

In [179]:
df_late_checkin = df[df['time_delta_with_previous_rental_in_minutes'] > 0]
df_late_checkin.describe()

rental_id         car_id  delay_at_checkout_in_minutes  \
count    1562.000000    1562.000000                   1288.000000   
mean   552354.698464  351153.017926                     24.229814   
std     12990.708193   55287.921131                    445.778648   
min    505560.000000  159533.000000                  -2676.000000   
25%    543791.250000  324307.750000                    -48.250000   
50%    552602.000000  367627.000000                      3.000000   
75%    562476.000000  391593.250000                     51.000000   
max    576292.000000  415953.000000                   9787.000000   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
count               1562.000000                                 1562.000000  
mean              550209.788732                                  329.174136  
std                13169.364374                                  244.891476  
min               505628.000000                                   30.000000  
25%               540903.000000                                  120.000000  
50%               550666.000000                                  240.000000  
75%               560786.750000                                  600.000000  
max               575053.000000                                  720.000000

How many problematic cases will it solve depending on the chosen threshold and scope?


In [152]:
df_prob = df[(df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['delay_at_checkout_in_minutes'] > 0)]


In [154]:
df_prob.state.value_counts()

state
ended    666
Name: count, dtype: int64

In [155]:
prob_cases = ((df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['delay_at_checkout_in_minutes']>0)).sum()
print(f"It would resolve", prob_cases, "problematic cases") 

print(f"This means {(100 * prob_cases) / df.shape[0]:.2f}% of cases over the total of rentals")

It would resolve 666 problematic cases
This means 3.13% of cases over the total of rentals


In [157]:
df_cancelled.previous_ended_rental_id.describe()

count       229.000000
mean     550913.327511
std       11955.397600
min      509972.000000
25%      543706.000000
50%      550970.000000
75%      560395.000000
max      574540.000000
Name: previous_ended_rental_id, dtype: float64

It would resolve 666 problematic cases.
This means 3.13% of cases over the total of rentals.
it is worth underlining that, the are not cancelled cars within this problematic cases, because none of the cancelled cars was reported as being check-out in delay. 

## How many problematic cases would it solve depending on the chosen threshold and scope?


Scope and threshold analysis

In [131]:
df.checkin_type.value_counts()

checkin_type
mobile     17003
connect     4307
Name: count, dtype: int64

In [132]:
# Assuming df contains the column 'checkin_type'
checkin_counts = df['checkin_type'].value_counts()

# Extracting data from value_counts()
total_checkin_counts = checkin_counts.sum()

percentages = (checkin_counts/total_checkin_counts) *100

In [30]:
labels = percentages.index.tolist()
values = percentages.values.tolist()




# Create the bar chart
fig = go.Figure(
    go.Bar(
        x=labels, 
        y=values,
        marker_color=['skyblue', 'lightcoral'],
        text=[f'{val:.2f}' for val in values],
        textposition='auto'
    )
)

# Customize the layout
fig.update_layout(
    title='Check-in Types Distribution',
    yaxis_title='Number of Check-ins',
    xaxis_title='Check-in Type'
)

# Display the chart
fig.show()


In [158]:
df_mobile = df[df['checkin_type']=='mobile']
df_mobile

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0         505000  363965       mobile  canceled                           NaN   
1         507750  269550       mobile     ended                         -81.0   
4         511440  313932       mobile     ended                           NaN   
5         511626  398802       mobile     ended                        -203.0   
7         512303  371242       mobile     ended                         -44.0   
...          ...     ...          ...       ...                           ...   
21302     573285  394863       mobile     ended                           NaN   
21303     573305  392752       mobile     ended                          72.0   
21305     573446  380069       mobile     ended                           NaN   
21306     573790  341965       mobile     ended                        -337.0   
21307     573791  364890       mobile     ended                         144.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                           NaN                                         NaN  
1                           NaN                                         NaN  
4                           NaN                                         NaN  
5                           NaN                                         NaN  
7                           NaN                                         NaN  
...                         ...                                         ...  
21302                       NaN                                         NaN  
21303                       NaN                                         NaN  
21305                  573429.0                                       300.0  
21306                       NaN                                         NaN  
21307                       NaN                                         NaN  

[17003 rows x 7 columns]

In [159]:
df_connect = df[df['checkin_type']=='connect']
df_connect


rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
2         508131  359049      connect     ended                          70.0   
3         508865  299063      connect  canceled                           NaN   
6         511639  370585      connect     ended                         -15.0   
9         513434  256528      connect     ended                          23.0   
11        514161  366037      connect  canceled                           NaN   
...          ...     ...          ...       ...                           ...   
21288     570001  386413      connect  canceled                           NaN   
21300     572293  355023      connect  canceled                           NaN   
21304     573322  376491      connect     ended                         -66.0   
21308     574852  362531      connect     ended                         -76.0   
21309     575056  351549      connect     ended                          35.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
2                           NaN                                         NaN  
3                           NaN                                         NaN  
6                      563782.0                                       570.0  
9                           NaN                                         NaN  
11                          NaN                                         NaN  
...                         ...                                         ...  
21288                       NaN                                         NaN  
21300                       NaN                                         NaN  
21304                       NaN                                         NaN  
21308                       NaN                                         NaN  
21309                       NaN                                         NaN  

[4307 rows x 7 columns]

In [160]:
df.time_delta_with_previous_rental_in_minutes.nlargest()

142     720.0
375     720.0
669     720.0
736     720.0
1175    720.0
Name: time_delta_with_previous_rental_in_minutes, dtype: float64

All types of cars

In [139]:

# Compute the number of rows where the time delta is between 0 and 30 minutes
num_cases_within_30_minutes_prob = ((df['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                (df['time_delta_with_previous_rental_in_minutes'] <= 30) & (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_30_minutes = 100 * num_cases_within_30_minutes_prob / prob_cases
# Display the result
print(f"Percentage of problematic cases within 30 minutes: {percentage_within_30_minutes:.2f}%")


# Calculate the percentage
percentage_within_30_minutes_over_total = 100 * num_cases_within_30_minutes_prob / df.shape[0]
# Display the result
print(f"Percentage of problematic cases within 30 minutes over the total of rentals: {percentage_within_30_minutes_over_total:.2f}%")

Percentage of problematic cases within 30 minutes: 8.56%
Percentage of problematic cases within 30 minutes over the total of rentals: 0.27%


In [161]:
((df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['time_delta_with_previous_rental_in_minutes']<=60)).sum()


# Compute the number of rows where the time delta is between 0 and 60 minutes
num_cases_within_60_minutes_prob = ( (df['time_delta_with_previous_rental_in_minutes'] > 0)& 
                               (df['time_delta_with_previous_rental_in_minutes']<=60) & (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_60_minutes = 100 * num_cases_within_60_minutes_prob / prob_cases
# Display the result
print(f"Percentage of problematic cases within 60 minutes: {percentage_within_60_minutes:.2f}%")


# Calculate the percentage
percentage_within_60_minutes_over_total = 100 * num_cases_within_60_minutes_prob / df.shape[0]
# Display the result
print(f"Percentage of problematic cases within 60 minutes over the total of rentals: {percentage_within_60_minutes_over_total:.2f}%")


Percentage of problematic cases within 60 minutes: 20.12%
Percentage of problematic cases within 60 minutes over the total of rentals: 0.63%


In [162]:
((df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['time_delta_with_previous_rental_in_minutes']<=120)).sum()




# Compute the number of rows where the time delta is between 0 and 120 minutes
num_cases_within_120_minutes_prob = ( (df['time_delta_with_previous_rental_in_minutes'] > 0)& 
                               (df['time_delta_with_previous_rental_in_minutes']<=120) & (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_120_minutes = 100 * num_cases_within_120_minutes_prob / prob_cases
# Display the result
print(f"Percentage of problematic cases within 120 minutes: {percentage_within_120_minutes:.2f}%")


# Calculate the percentage
percentage_within_120_minutes_over_total = 100 * num_cases_within_120_minutes_prob / df.shape[0]
# Display the result
print(f"Percentage of problematic cases within 120 minutes over the total of rentals: {percentage_within_120_minutes_over_total:.2f}%")

Percentage of problematic cases within 120 minutes: 35.89%
Percentage of problematic cases within 120 minutes over the total of rentals: 1.12%


In [163]:
((df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['time_delta_with_previous_rental_in_minutes']<=240)).sum()



# Compute the number of rows where the time delta is between 0 and 240 minutes
num_cases_within_240_minutes_prob = ( (df['time_delta_with_previous_rental_in_minutes'] > 0)& 
                               (df['time_delta_with_previous_rental_in_minutes']<=240) &  (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_240_minutes = 100 * num_cases_within_240_minutes_prob / prob_cases
# Display the result
print(f"Percentage of problematic cases within 240 minutes: {percentage_within_240_minutes:.2f}%")

# Calculate the percentage
percentage_within_240_minutes_over_total = 100 * num_cases_within_240_minutes_prob / df.shape[0]
# Display the result
print(f"Percentage of problematic cases within 240 minutes over the total of rentals: {percentage_within_240_minutes_over_total:.2f}%")

Percentage of problematic cases within 240 minutes: 53.30%
Percentage of problematic cases within 240 minutes over the total of rentals: 1.67%


In [164]:
((df['time_delta_with_previous_rental_in_minutes'] > 0) & (df['time_delta_with_previous_rental_in_minutes']<=600)).sum()


# Compute the number of rows where the time delta is between 0 and 600 minutes
num_cases_within_600_minutes_prob = ((df['time_delta_with_previous_rental_in_minutes'] > 0)& 
                               (df['time_delta_with_previous_rental_in_minutes']<=600)& (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_600_minutes = 100 * num_cases_within_600_minutes_prob / prob_cases
# Display the result
print(f"Percentage of problematic cases within 600 minutes: {percentage_within_600_minutes:.2f}%")


# Calculate the percentage
percentage_within_600_minutes_over_total = 100 * num_cases_within_600_minutes_prob / df.shape[0]
# Display the result
print(f"Percentage of problematic cases within 600 minutes over the total of rentals: {percentage_within_600_minutes_over_total:.2f}%")

Percentage of problematic cases within 600 minutes: 81.23%
Percentage of problematic cases within 600 minutes over the total of rentals: 2.54%


In [169]:

# Compute the number of rows where the time delta is between 0 and 600 minutes
num_cases_within_720_minutes_problem = ( (df['time_delta_with_previous_rental_in_minutes'] > 0)& 
                               (df['time_delta_with_previous_rental_in_minutes']<=720) & (df['delay_at_checkout_in_minutes']>0)).sum()

# Calculate the percentage
percentage_within_720_minutes = 100 * (num_cases_within_720_minutes_problem / prob_cases)
# Display the result
print(f"Percentage of problematic cases within 720 minutes: {percentage_within_720_minutes:.2f}%")

total_cases_within_720_minutes = ((df['time_delta_with_previous_rental_in_minutes'] > 0) & 
                               (df['time_delta_with_previous_rental_in_minutes']<=720)).sum()
percentage_prob_case_within_720_minutes = 100* (num_cases_within_720_minutes_problem/total_cases_within_720_minutes)

print(f"Percentage of the total of problematic cases within 720 delta between rentals: {percentage_prob_case_within_720_minutes:.2f}%")
# Calculate the percentage
percentage_within_720_minutes_over_total = 100 * num_cases_within_720_minutes_problem / df.shape[0]
# Display the result    
print(f"Percentage of problematic cases within 720 minutes over the total of rentals: {percentage_within_720_minutes_over_total:.2f}%")




percentage_total_cases_within_720_minutes = 100 * total_cases_within_720_minutes / df.shape[0]
# Display the result    
print(f"Total rentals within 720 minutes over the total of rentals: {percentage_total_cases_within_720_minutes:.2f}%")

Percentage of problematic cases within 720 minutes: 100.00%
Percentage of the total of problematic cases within 720 delta between rentals: 42.64%
Percentage of problematic cases within 720 minutes over the total of rentals: 3.13%
Total rentals within 720 minutes over the total of rentals: 7.33%


In [167]:
total_cases_within_720_minutes

1562

In [166]:
num_cases_within_720_minutes_problem

666

## If a threshold imposing of at list 720 minutes (12 hours) of differences between one rental and another is applied to all cars this would concern 7.33% of the total of rental where the delta between two rentals is less than 12 hours, 
# Within this threshold of 720 minutes such a feature would the 42,6% of the problematic cases where in a delay in the following rental was verified.

In [170]:
df[(df['time_delta_with_previous_rental_in_minutes'] <= 0) & (df['delay_at_checkout_in_minutes']>0)]

rental_id  car_id checkin_type  state  delay_at_checkout_in_minutes  \
107       537576  397470       mobile  ended                          18.0   
148       540479  374684       mobile  ended                          12.0   
164       541862  382364       mobile  ended                         125.0   
342       552330  384657      connect  ended                          23.0   
438       559781  408776       mobile  ended                          44.0   
...          ...     ...          ...    ...                           ...   
20777     552034  361156       mobile  ended                         335.0   
20802     553351  406124      connect  ended                          90.0   
20924     561476  410402       mobile  ended                          11.0   
21054     562649  379751      connect  ended                          72.0   
21163     565721  381470       mobile  ended                          44.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
107                    539005.0                                         0.0  
148                    539751.0                                         0.0  
164                    540607.0                                         0.0  
342                    554800.0                                         0.0  
438                    540868.0                                         0.0  
...                         ...                                         ...  
20777                  548172.0                                         0.0  
20802                  557210.0                                         0.0  
20924                  550186.0                                         0.0  
21054                  565386.0                                         0.0  
21163                  564855.0                                         0.0  

[136 rows x 7 columns]

Connect cars

In [83]:
# Compute the number of rows where the time delta is between 0 and 30 minutes
num_cases_within_30_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                               (df_connect['time_delta_with_previous_rental_in_minutes'] <= 30) & 
                               (df_connect['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_30_minutes_connect = 100 * num_cases_within_30_minutes_connect / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 30 minutes: {percentage_within_30_minutes_connect:.2f}%")

# Calculate the percentage of cases within 30 minutes over the total number of rentals
percentage_within_30_minutes_over_total_connect = 100 * num_cases_within_30_minutes_connect / df_connect.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 30 minutes over the total of rentals: {percentage_within_30_minutes_over_total_connect:.2f}%")

# Compute the number of rows where the time delta is between 0 and 60 minutes
num_cases_within_60_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                               (df_connect['time_delta_with_previous_rental_in_minutes'] <= 60) & 
                               (df_connect['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_60_minutes_connect = 100 * num_cases_within_60_minutes_connect / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 60 minutes: {percentage_within_60_minutes_connect:.2f}%")

# Calculate the percentage of cases within 60 minutes over the total number of rentals
percentage_within_60_minutes_over_total_connect = 100 * num_cases_within_60_minutes_connect / df_connect.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 60 minutes over the total of rentals: {percentage_within_60_minutes_over_total_connect:.2f}%")

# Compute the number of rows where the time delta is between 0 and 120 minutes
num_cases_within_120_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                (df_connect['time_delta_with_previous_rental_in_minutes'] <= 120) & 
                                (df_connect['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_120_minutes_connect = 100 * num_cases_within_120_minutes_connect / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 120 minutes: {percentage_within_120_minutes_connect:.2f}%")

# Calculate the percentage of cases within 120 minutes over the total number of rentals
percentage_within_120_minutes_over_total_connect = 100 * num_cases_within_120_minutes_connect / df_connect.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 120 minutes over the total of rentals: {percentage_within_120_minutes_over_total_connect:.2f}%")

# Compute the number of rows where the time delta is between 0 and 240 minutes
num_cases_within_240_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                (df_connect['time_delta_with_previous_rental_in_minutes'] <= 240) & 
                                (df_connect['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_240_minutes_connect = 100 * num_cases_within_240_minutes_connect / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 240 minutes: {percentage_within_240_minutes_connect:.2f}%")

# Calculate the percentage of cases within 240 minutes over the total number of rentals
percentage_within_240_minutes_over_total_connect = 100 * num_cases_within_240_minutes_connect / df_connect.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 240 minutes over the total of rentals: {percentage_within_240_minutes_over_total_connect:.2f}%")



num_cases_within_600_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                (df_connect['time_delta_with_previous_rental_in_minutes'] <= 600) & (df['delay_at_checkout_in_minutes']>0) ).sum()

# Calculate the percentage
percentage_within_600_minutes_connect  = 100 * num_cases_within_600_minutes_connect  / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 600 minutes: {percentage_within_600_minutes_connect:.2f}%")

# Calculate the percentage of cases within 600 minutes over the total number of rentals
percentage_within_600_minutes_over_total_connect = 100 * num_cases_within_600_minutes_connect / df.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 600 minutes over the total of rentals: {percentage_within_600_minutes_over_total_connect:.2f}%")

# Compute the number of rows where the time delta is between 0 and 600 minutes
num_cases_within_720_minutes_connect = ((df_connect['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                (df_connect['time_delta_with_previous_rental_in_minutes'] <= 720) & (df['delay_at_checkout_in_minutes']>0)).sum()
_
percentage_within_720_minutes_connect = 100 * num_cases_within_720_minutes_connect / prob_cases
# Display the result
print(f"Percentage of problematic connect cases within 720 minutes: {percentage_within_720_minutes_connect:.2f}%")

# Calculate the percentage of cases within 720 minutes over the total number of rentals
percentage_within_720_minutes_over_total_connect = 100 * num_cases_within_720_minutes_connect / df.shape[0]
# Display the result
print(f"Percentage of problematic connect cases within 720 minutes over the total of rentals: {percentage_within_720_minutes_over_total_connect:.2f}%")



Percentage of problematic connect cases within 30 minutes: 2.85%
Percentage of problematic connect cases within 30 minutes over the total of rentals: 0.44%
Percentage of problematic connect cases within 60 minutes: 6.91%
Percentage of problematic connect cases within 60 minutes over the total of rentals: 1.07%
Percentage of problematic connect cases within 120 minutes: 11.56%
Percentage of problematic connect cases within 120 minutes over the total of rentals: 1.79%
Percentage of problematic connect cases within 240 minutes: 17.27%
Percentage of problematic connect cases within 240 minutes over the total of rentals: 2.67%
Percentage of problematic connect cases within 600 minutes: 26.73%
Percentage of problematic connect cases within 600 minutes over the total of rentals: 0.84%
Percentage of problematic connect cases within 720 minutes: 33.63%
Percentage of problematic connect cases within 720 minutes over the total of rentals: 1.05%


Mobile cars

In [49]:
# Compute the number of rows where the time delta is between 0 and 30 minutes
num_cases_within_30_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                       (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 30) & 
                                       (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_30_minutes_mobile = 100 * num_cases_within_30_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 30 minutes: {percentage_within_30_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 30 minutes over the total number of rentals
percentage_within_30_minutes_over_total_mobile = 100 * num_cases_within_30_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 30 minutes over the total of rentals: {percentage_within_30_minutes_over_total_mobile:.2f}%")

# Compute the number of rows where the time delta is between 0 and 60 minutes
num_cases_within_60_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                       (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 60) & 
                                       (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_60_minutes_mobile = 100 * num_cases_within_60_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 60 minutes: {percentage_within_60_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 60 minutes over the total number of rentals
percentage_within_60_minutes_over_total_mobile = 100 * num_cases_within_60_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 60 minutes over the total of rentals: {percentage_within_60_minutes_over_total_mobile:.2f}%")

# Compute the number of rows where the time delta is between 0 and 120 minutes
num_cases_within_120_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                        (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 120) & 
                                        (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_120_minutes_mobile = 100 * num_cases_within_120_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 120 minutes: {percentage_within_120_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 120 minutes over the total number of rentals
percentage_within_120_minutes_over_total_mobile = 100 * num_cases_within_120_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 120 minutes over the total of rentals: {percentage_within_120_minutes_over_total_mobile:.2f}%")

# Compute the number of rows where the time delta is between 0 and 240 minutes
num_cases_within_240_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                        (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 240) & 
                                        (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_240_minutes_mobile = 100 * num_cases_within_240_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 240 minutes: {percentage_within_240_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 240 minutes over the total number of rentals
percentage_within_240_minutes_over_total_mobile = 100 * num_cases_within_240_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 240 minutes over the total of rentals: {percentage_within_240_minutes_over_total_mobile:.2f}%")

# Compute the number of rows where the time delta is between 0 and 600 minutes
num_cases_within_600_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                        (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 600) & 
                                        (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_600_minutes_mobile = 100 * num_cases_within_600_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 600 minutes: {percentage_within_600_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 600 minutes over the total number of rentals
percentage_within_600_minutes_over_total_mobile = 100 * num_cases_within_600_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 600 minutes over the total of rentals: {percentage_within_600_minutes_over_total_mobile:.2f}%")

# Compute the number of rows where the time delta is between 0 and 720 minutes
num_cases_within_720_minutes_mobile = ((df_mobile['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                        (df_mobile['time_delta_with_previous_rental_in_minutes'] <= 720) & 
                                        (df_mobile['delay_at_checkout_in_minutes'] > 0)).sum()

# Calculate the percentage
percentage_within_720_minutes_mobile = 100 * num_cases_within_720_minutes_mobile / prob_cases
# Display the result
print(f"Percentage of problematic mobile cases within 720 minutes: {percentage_within_720_minutes_mobile:.2f}%")

# Calculate the percentage of cases within 720 minutes over the total number of rentals
percentage_within_720_minutes_over_total_mobile = 100 * num_cases_within_720_minutes_mobile / df.shape[0]
# Display the result
print(f"Percentage of problematic mobile cases within 720 minutes over the total of rentals: {percentage_within_720_minutes_over_total_mobile:.2f}%")


Percentage of problematic mobile cases within 30 minutes: 5.71%
Percentage of problematic mobile cases within 30 minutes over the total of rentals: 0.18%
Percentage of problematic mobile cases within 60 minutes: 13.21%
Percentage of problematic mobile cases within 60 minutes over the total of rentals: 0.41%
Percentage of problematic mobile cases within 120 minutes: 24.32%
Percentage of problematic mobile cases within 120 minutes over the total of rentals: 0.76%
Percentage of problematic mobile cases within 240 minutes: 36.04%
Percentage of problematic mobile cases within 240 minutes over the total of rentals: 1.13%
Percentage of problematic mobile cases within 600 minutes: 54.50%
Percentage of problematic mobile cases within 600 minutes over the total of rentals: 1.70%
Percentage of problematic mobile cases within 720 minutes: 66.37%
Percentage of problematic mobile cases within 720 minutes over the total of rentals: 2.07%


In [56]:
import pandas as pd

def calculate_percentages(df, prob_cases, time_intervals, dataset_name):
    for interval in time_intervals:
        # Compute the number of rows where the time delta is between 0 and the current interval
        num_cases_within_interval = ((df['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                      (df['time_delta_with_previous_rental_in_minutes'] <= interval) & 
                                      (df['delay_at_checkout_in_minutes'] > 0)).sum()

        # Calculate the percentage within the interval
        percentage_within_interval = 100 * num_cases_within_interval / prob_cases
        print(f"Percentage of problematic {dataset_name} cases within {interval} minutes: {percentage_within_interval:.2f}%")

        # Calculate the percentage of cases within the interval over the total number of rentals
        percentage_within_interval_over_total = 100 * num_cases_within_interval / df.shape[0]
        print(f"Percentage of problematic {dataset_name} cases within {interval} minutes over the total of rentals: {percentage_within_interval_over_total:.2f}%")

# Define the time intervals
time_intervals = [30, 60, 120, 240, 600, 720]

# Sample data for df_connect and df_mobile
# df_connect = pd.DataFrame({'time_delta_with_previous_rental_in_minutes': [...], 'delay_at_checkout_in_minutes': [...]})
# df_mobile = pd.DataFrame({'time_delta_with_previous_rental_in_minutes': [...], 'delay_at_checkout_in_minutes': [...]})
# prob_cases = ... # Define the number of problematic cases

# Calculate and print percentages for df_connect
calculate_percentages(df_connect, prob_cases, time_intervals, 'connect')

# Calculate and print percentages for df_mobile
calculate_percentages(df_mobile, prob_cases, time_intervals, 'mobile')


Percentage of problematic connect cases within 30 minutes: 2.85%
Percentage of problematic connect cases within 30 minutes over the total of rentals: 0.44%
Percentage of problematic connect cases within 60 minutes: 6.91%
Percentage of problematic connect cases within 60 minutes over the total of rentals: 1.07%
Percentage of problematic connect cases within 120 minutes: 11.56%
Percentage of problematic connect cases within 120 minutes over the total of rentals: 1.79%
Percentage of problematic connect cases within 240 minutes: 17.27%
Percentage of problematic connect cases within 240 minutes over the total of rentals: 2.67%
Percentage of problematic connect cases within 600 minutes: 26.73%
Percentage of problematic connect cases within 600 minutes over the total of rentals: 4.13%
Percentage of problematic connect cases within 720 minutes: 33.63%
Percentage of problematic connect cases within 720 minutes over the total of rentals: 5.20%
Percentage of problematic mobile cases within 30 min

In [50]:
# Data for connect and mobile cars
time_intervals = ['0-30', '0-60', '0-120', '0-240', '0-600', '0-720']

# Example percentages for connect and mobile cases (replace with your actual data)
percentages_connect = [
    percentage_within_30_minutes_connect, 
    percentage_within_60_minutes_connect, 
    percentage_within_120_minutes_connect, 
    percentage_within_240_minutes_connect, 
    percentage_within_600_minutes_connect,
    percentage_within_720_minutes_connect
]

percentages_mobile = [
    percentage_within_30_minutes_mobile, 
    percentage_within_60_minutes_mobile, 
    percentage_within_120_minutes_mobile, 
    percentage_within_240_minutes_mobile, 
    percentage_within_600_minutes_mobile,
    percentage_within_720_minutes_mobile
]

# Create Plotly figure
fig = go.Figure()

# Add bars for connect check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_connect,
    name='Connect Check-in',
    marker_color='lightcoral'
))

# Add bars for mobile check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_mobile,
    name='Mobile Check-in',
    marker_color='royalblue'
))

# Update the layout
fig.update_layout(
    title='Percentage over total of problematic cases for the different time intervals',
    xaxis_title='Time Interval (minutes)',
    yaxis_title='Percentage (%)',
    barmode='group',  # Group the bars
    xaxis=dict(tickmode='array', tickvals=time_intervals),
    yaxis=dict(range=[0, max(max(percentages_connect), max(percentages_mobile)) + 10])
)

# Show the chart in Streamlit
fig.show()

In [69]:
import streamlit as st

def calculate_percentages(df, prob_cases, time_intervals, dataset_name):
    percentages_within_intervals = []
    percentages_within_intervals_over_total = []
    
    for interval in time_intervals:
        # Compute the number of rows where the time delta is between 0 and the current interval
        num_cases_within_interval = ((df['time_delta_with_previous_rental_in_minutes'] > 0) & 
                                      (df['time_delta_with_previous_rental_in_minutes'] <= interval) & 
                                      (df['delay_at_checkout_in_minutes'] > 0)).sum()

        # Calculate the percentage within the interval
        percentage_within_interval = 100 * num_cases_within_interval / prob_cases
        percentages_within_intervals.append(percentage_within_interval)

        # Calculate the percentage of cases within the interval over the total number of rentals
        percentage_within_interval_over_total = 100 * num_cases_within_interval / df.shape[0]
        percentages_within_intervals_over_total.append(percentage_within_interval_over_total)
    
    return percentages_within_intervals, percentages_within_intervals_over_total


In [171]:
# Assuming `df_connect` and `df_mobile` are your dataframes, and `prob_cases_connect`, `prob_cases_mobile` are the total problematic cases

# Calculate percentages for both datasets
percentages_connect, percentages_connect_over_total = calculate_percentages(df_connect, prob_cases, time_intervals, 'connect')
percentages_mobile, percentages_mobile_over_total = calculate_percentages(df_mobile, prob_cases, time_intervals, 'mobile')


In [76]:
# Create Plotly figure
fig = go.Figure()

# Add bars for connect check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_connect,
    name='Connect Check-in',
    marker_color='lightcoral'
))

# Add bars for mobile check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_mobile,
    name='Mobile Check-in',
    marker_color='royalblue'
))

# Update the layout
fig.update_layout(
    title='Percentage over problematic cases for the different time intervals',
    xaxis_title='Time Interval (minutes)',
    yaxis_title='Percentage (%)',
    barmode='group',  # Group the bars
    xaxis=dict(tickmode='array', tickvals=time_intervals),
    yaxis=dict(range=[0, max(max(percentages_connect), max(percentages_mobile_over_total)) + 10])
)

# Show the chart in Streamlit
fig.show()

In [68]:
df.checkin_type.value_counts()

checkin_type
mobile     17003
connect     4307
Name: count, dtype: int64

In [80]:
# Create Plotly figure
fig = go.Figure()

# Add bars for connect check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_connect_over_total,
    name='Connect Check-in',
    marker_color='lightcoral'
))

# Add bars for mobile check-in cases
fig.add_trace(go.Bar(
    x=time_intervals,
    y=percentages_mobile_over_total,
    name='Mobile Check-in',
    marker_color='royalblue'
))

# Update the layout
fig.update_layout(
    title='Percentage over total rented cars for the different time intervals',
    xaxis_title='Time Interval (minutes)',
    yaxis_title='Percentage (%)',
    barmode='group',  # Group the bars
    xaxis=dict(tickmode='array', tickvals=time_intervals),
)

# Show the chart in Streamlit
fig.show()

In [81]:
percentages_mobile_over_total

[0.2234899723578192,
 0.5175557254602129,
 0.9527730400517556,
 1.4115156148914898,
 2.134917367523378,
 2.59954125742516]

In [82]:
percentages_connect_over_total

[0.4411423264453216,
 1.0680287903413048,
 1.7877873229626189,
 2.6700719758532623,
 4.132807058277223,
 5.200835848618528]

## If the feature's scope applies only ofr connect car with a maximum threshold time of 720 minutes (12 hours), which is the longest measured, this may have an impact on 1.05% of the total rentals where problematic situations were verified. 

## If the feature's scope applies both to mobile or connect cars with a maximum threshold time of 720 minutes (12 hours), it may affect 2.07% of the total rentals where problematic situations were verified. 

